onnx_caffe2/backend.py

```diff
@@ -103,10 +103,14 @@ def run(self, inputs, **kwargs):
                         for key, value in inputs.items():
                             workspace.FeedBlob(key, value)
                 elif isinstance(inputs, list) or isinstance(inputs, tuple):
-                    assert len(self.uninitialized) == len(inputs)
-                    for i, value in enumerate(inputs):
-                        # namescope already baked into protobuf
-                        workspace.FeedBlob(self.uninitialized[i], value)
+                    if len(self.uninitialized) == len(inputs):
```

Comment:
How about we don't store `uninitialized` in `Caffe2Rep` and do it this way:

```
i = 0
for input_name in predict_net.external_input:
    if workspace.HasBlob(input_name):
        continue
    if i > len(inputs):
        raise RuntimeError(
            "Not enough inputs, unable to initialize {}".format(input_name))
    workspace.FeedBlob(input_name, inputs[i])
    i += 1
```

This way we can handle the case that user runs `init_net` to fill in some blobs after the predict_net has been prepared.

Reply:
One thing is that we do need to store uninitialized because we want to prepare once and run multiple times. But I think the current code still assumes some order of input blobs, so I'll change the first part to something similar to this. 
This pr is actually to handle the case where some models has "input" already filled even before we give the actual input to the network.

Reply:
Actually, we have no way to do it except for assuming the order of input matches the order of uninitialized blobs when user feed a list of input blob to "run" function. So I think we should encourage people to use list/tuple input only when there is one input, for other cases, people should use the dictionary input.

Reply:
On a second thought, I think the existing implementation is better. `prepare` was kinda like currying: once prepared, the unbound (which is `uninitialized` in our case) inputs are fixed and users are supposed to provide all of them in `run`.

btw, I think we don't need to store `init_net` in `Caffe2Rep`

Reply:
"This pr is actually to handle the case where some models has "input" already filled even before we give the actual input to the network."

You mean `init_graph` in the `prepare` function? But the blobs initialized by init_graph should have been excluded from `uninitialized`:
```
uninitialized = [x
                 for x in predict_net.external_input
                 if not workspace.HasBlob(x)]
```
right?

Reply:
> This pr is actually to handle the case where some models has "input" already filled even before we give the actual input to the network.

It would be really helpful if we could see the sample code that triggers this situation. It's hard to evaluate this proposal without knowing what it solves.